doc/design/refactor/multi_device.md

```diff
@@ -0,0 +1,37 @@
+# Design Doc: Execute the Program with Multi Thread
+
+## Abstract
+
+This Design Doc propose an approach to make the user-defined Op graph
+running with multi threads, we will use an auto transpiler to convert the user-defined
+Op graph to a multi-thread Op graph, and run `ParallelDo` Op to run the
+multi-thread graph.
+
+## Graph Converter
+<img src="src/multi-threads/single-thread@3x.png" width="300">
+
+After converted:
+
+<img src="src/multi-threads/multi-threads@3x.png" width="1000">
```

Comment:
Can this figure also show the scopes these blocks use? And when how to ensure thread-safety when doing the merge?

Reply:
I think the `Merge` Op should waiting for all forward/backward Ops completed.